formatters/cloudevents/formatter.go

```diff
@@ -0,0 +1,188 @@
+	// combination of Source + ID must be unique.  Events with the same Source +
+	// ID can be assumed to be duplicates by consumers
+	ID string `json:"id"`
+
+	// Source identifies the context in which the event happened, it is a
+	// URI-reference, cannot be empty and is required.
+	Source string `json:"source"`
+
+	// SpecVersion defines the version of CloudEvents that the event is using,
+	// it cannot be empty and is required.
+	SpecVersion string `json:"specversion"`
+
+	// Type defines the event's type, cannot be empty and is required.
+	Type string `json:"type"`
+
+	// Data may include domain-specific information about the event and is
+	// optional.
+	Data interface{} `json:"data,omitempty"`
+
+	// DataContentType defines the content type of the event's data value and is
+	// optional.  If present it must adhere to:
+	// https://datatracker.ietf.org/doc/html/rfc2046
+	DataContentType string `json:"datacontentype,omitempty"`
+
+	// DataSchema is a URI-reference and is optional.
+	DataSchema string `json:"dataschema,omitempty"`
+
+	// Time is in format RFC 3339 (the default for time.Time) and is optional
+	Time time.Time `json:"time,omitempty"`
+}
+
+// Formatter is a Node which formats the Event as a CloudEvent in JSON
+// format (See: https://github.com/cloudevents/spec)
+type Formatter struct {
+	// Source identifies the context where the cloudevents happen and is
+	// required
+	Source *url.URL
+
+	// Schema is the JSON schema for the cloudevent data and is optional
+	Schema *url.URL
+
+	// Format defines the format created by the node.  If empty (unspecified),
+	// FormatJSON will be used
+	Format Format
+}
+
+var _ eventlogger.Node = &Formatter{}
+
+func (f *Formatter) validate() error {
+	const op = "cloudevents.(Formatter).validate"
+	if f == nil {
+		return fmt.Errorf("%s: missing formatter: %w", op, eventlogger.ErrInvalidParameter)
+	}
+	if f.Source == nil || f.Source.String() == "" {
+		return fmt.Errorf("%s: missing source: %w", op, eventlogger.ErrInvalidParameter)
+	}
+	if err := f.Format.validate(); err != nil {
+		return fmt.Errorf("%s: %w", op, err)
+	}
+	if f.Schema != nil && f.Schema.String() == "" {
+		return fmt.Errorf("%s: an empty schema is not valid: %w", op, eventlogger.ErrInvalidParameter)
+	}
+	return nil
+}
+
+// Process formats the Event as a cloudevent and stores that formatted data in
+// Event.Formatted with a key of "cloudevents-json" (cloudevents.FormatJSON)
+func (f *Formatter) Process(ctx context.Context, e *eventlogger.Event) (*eventlogger.Event, error) {
+	const op = "cloudevents.(Formatter).Process"
+	if err := f.validate(); err != nil {
+		return nil, fmt.Errorf("%s: invalid Formatter %w", op, err)
+	}
+	if e == nil {
+		return nil, fmt.Errorf("%s: missing event: %w", op, eventlogger.ErrInvalidParameter)
+	}
+
+	var data interface{}
+	if i, ok := e.Payload.(Data); ok {
+		data = i.Data()
+	} else {
+		data = e.Payload
+	}
+	var id string
+	if i, ok := e.Payload.(ID); ok {
+		id = i.ID()
```

Comment:
Should we error if `ID()` returns an empty string here?

Reply:
yes. ty!